File: ios/Anemone SDK/general/json/JSONObject.swift
//
//  JSONObject.swift
//  Anemone SDK
//
//  Created by Angel Garcia on 17/01/2017.
//  Copyright © 2017 Syltek Solutions S.L. All rights reserved.
//

import Foundation

public class JSONObject {
    var data: [String: Any]

    public init() {
        self.data = [:]
    }

    init(data: [String: Any]) {
        self.data = data
    }

    public convenience init(string: String) throws {
        guard let data = string.data(using: .utf8) else {
            throw AnemoneException.jsonInvalidFormat(key: nil)
        }
        try self.init(data: data)
    }

    public convenience init(data: Data) throws {
        guard let json = try JSONSerialization.jsonObject(with: data, options: []) as? [String: Any] else {
            throw AnemoneException.jsonInvalidFormat(key: nil)
        }
        self.init(data: json)
    }

    public func has(_ key: String) -> Bool {
        data[key] != nil && !(data[key] is NSNull)
    }

    public func keys() -> [String] {
        data.keys.map { $0 }
    }

    public func getAny(_ key: String) throws -> Any {
        guard let value = data[key], !(value is NSNull) else {
            throw AnemoneException.jsonInvalidFormat(key: key)
        }

        return value
    }

    public func optAny(_ key: String) -> Any? {
        try? getAny(key)
    }

    public func getInt(_ key: String) throws -> Int {
        guard let value = data[key] as? Int else {
            throw AnemoneException.jsonInvalidFormat(key: key)
        }
        return value
    }

    public func optInt(_ key: String) -> Int? {
        try? getInt(key)
    }

    public func getDouble(_ key: String) throws -> Double {
        guard let value = data[key] as? Double else {
            throw AnemoneException.jsonInvalidFormat(key: key)
        }
        return value
    }

    public func optDouble(_ key: String) -> Double? {
        try? getDouble(key)
    }

    public func getDecimal(_ key: String, decimals: Int) throws -> Decimal {
        if let value = data[key] as? Double {
            return Decimal(value).roundedValue(decimals: decimals)
        } else if let valueInt = data[key] as? Int {
            return Decimal(Double(valueInt)).roundedValue(decimals: decimals)
        }
        throw AnemoneException.jsonInvalidFormat(key: key)
    }

    public func optDecimal(_ key: String, decimals: Int) -> Decimal? {
        try? getDecimal(key, decimals: decimals)
    }

    public func getBoolean(_ key: String) throws -> Bool {
        guard let value = data[key] as? Bool else {
            throw AnemoneException.jsonInvalidFormat(key: key)
        }
        return value
    }

    public func optBoolean(_ key: String) -> Bool? {
        try? getBoolean(key)
    }

    public func getString(_ key: String) throws -> String {
        guard let value = data[key] as? String else {
            throw AnemoneException.jsonInvalidFormat(key: key)
        }
        return value
    }

    public func optString(_ key: String) -> String? {
        try? getString(key)
    }

    public func getDate(_ key: String) throws -> Date {
        guard let value = data[key] as? String else {
            throw AnemoneException.jsonInvalidFormat(key: key)
        }
        guard let date = Date.from(string: value, format: "\(Date.defaultFormat)'Z'", timeZone: TimeZone.utc) else {
            throw AnemoneException.jsonInvalidFormat(key: key)
        }
        return date
    }

    public func optDate(_ key: String) -> Date? {
        try? getDate(key)
    }

    public func getTime(_ key: String) throws -> Time {
        guard let value = data[key] as? String else {
            throw AnemoneException.jsonInvalidFormat(key: key)
        }
        let components = value.split(separator: ":").compactMap { Int($0) }
        guard components.count == 3 else {
            throw AnemoneException.jsonInvalidFormat(key: key)
        }
        return Time.of(hour: components[0], minute: components[1], second: components[2])
    }

    public func optTime(_ key: String) -> Time? {
        try? getTime(key)
    }

    public func getJSONObject(_ key: String) throws -> JSONObject {
        guard let value = data[key] as? [String: Any] else {
            throw AnemoneException.jsonInvalidFormat(key: key)
        }
        return JSONObject(data: value)
    }

    public func optJSONObject(_ key: String) -> JSONObject? {
        try? getJSONObject(key)
    }

    public func getJSONArray(_ key: String) throws -> JSONArray {
        guard let value = data[key] as? [Any] else {
            throw AnemoneException.jsonInvalidFormat(key: key)
        }
        return JSONArray(data: value)
    }

    public func optJSONArray(_ key: String) -> JSONArray? {
        try? getJSONArray(key)
    }

    public func getStringArray(_ key: String) throws -> [String] {
        guard let value = data[key] as? [String] else {
            throw AnemoneException.jsonInvalidFormat(key: key)
        }
        return value
    }

    public func optStringArray(_ key: String) -> [String]? {
        try? getStringArray(key)
    }

    public func getIntArray(_ key: String) throws -> [Int] {
        guard let value = data[key] as? [Int] else {
            throw AnemoneException.jsonInvalidFormat(key: key)
        }
        return value
    }

    public func getAnyArray(_ key: String) throws -> [Any] {
        guard let value = data[key] as? [Any] else {
            throw AnemoneException.jsonInvalidFormat(key: key)
        }
        return value
    }

    public func asMap() -> [String: Any] {
        data.filter { _, value in !(value is NSNull) }
    }

    public func setAny(_ key: String, _ value: Any?) {
        data[key] = value
    }

    public func setInt(_ key: String, _ value: Int?) {
        data[key] = value
    }

    public func setDouble(_ key: String, _ value: Double?) {
        data[key] = value
    }

    public func setDecimal(_ key: String, _ value: Decimal?) {
        data[key] = value
    }

    public func setString(_ key: String, _ value: String?) {
        data[key] = value
    }

    public func setBoolean(_ key: String, _ value: Bool?) {
        data[key] = value
    }

    public func setObject(_ key: String, _ value: JSONObject?) {
        data[key] = value?.data
    }

    public func setArray(_ key: String, _ value: [Any]?) {
        data[key] = value
    }

    public func setJSONArray(_ key: String, _ value: JSONArray?) {
        data[key] = value?.data
    }

    public func setStringArray(_ key: String, _ value: [String]?) {
        data[key] = value
    }
    
    public func setIntArray(_ key: String, _ value: [Int]?) {
        data[key] = value
    }

    public func setDate(_ key: String, _ value: Date?) {
        guard let value = value else {
            data[key] = nil
            return
        }

        data[key] = value.toString(format: Date.defaultFormat, timeZone: TimeZone.utc)
    }

    public func setTime(_ key: String, _ value: Time?) {
        guard let value = value else {
            data[key] = nil
            return
        }
        data[key] = value.description
    }

    public func toData() throws -> Data {
        guard let jsonData = try? JSONSerialization.data(withJSONObject: data, options: []) else {
            throw AnemoneException.jsonInvalidFormat(key: nil)
        }
        return jsonData
    }

    public func toString() throws -> String {
        let jsonData = try toData()
        guard let jsonString = String(data: jsonData, encoding: .utf8) else {
            throw AnemoneException.jsonInvalidFormat(key: nil)
        }
        return jsonString.replacingOccurrences(of: "\\/", with: "/")
    }
}

File: android/app/src/main/java/com/playtomic/test/anemone_sdk/general/json/JSONObject.kt
@file:Suppress("UNCHECKED_CAST")

package com.anemonesdk.general.json

import com.playtomic.foundation.extension.Dates
import com.playtomic.foundation.extension.compactMap
import com.playtomic.foundation.extension.roundedValue
import com.playtomic.foundation.extension.toString
import com.playtomic.foundation.model.Time
import org.json.JSONException
import java.math.BigDecimal
import java.util.*

/**
 * Created by mgonzalez on 18/1/17.
 */

class JSONObject(internal val data: org.json.JSONObject = org.json.JSONObject()) {

    constructor(string: String) : this(data = org.json.JSONObject(string))

    constructor(data: ByteArray) : this(string = String(data))

    constructor(data: Map<String, Any>) : this() {
        data.forEach { (key, value) ->
            setAny(key, value)
        }
    }

    fun has(name: String) = !data.isNull(name)

    fun keys(): List<String> = data.keys().asSequence().toList()

    fun optAny(name: String): Any? =
        if (data.isNull(name)) null else data.opt(name)

    @Throws(JSONException::class)
    fun getAny(name: String) = optAny(name) ?: throw JSONException("Expected value for " + name)

    fun optInt(name: String): Int? {
        val obj = data.opt(name)
        if (obj is Int) {
            return obj
        }
        return null
    }

    @Throws(JSONException::class)
    fun getInt(name: String) = optInt(name) ?: throw JSONException("Expected int for " + name)

    fun optDouble(name: String): Double? {
        val obj = data.opt(name)
        if (obj is Double) {
            return obj
        } else if (obj is Number) {
            return obj.toDouble()
        }

        return null
    }

    @Throws(JSONException::class)
    fun getDouble(name: String) = optDouble(name)
        ?: throw JSONException("Expected double for " + name)

    fun optDecimal(name: String, decimals: Int): BigDecimal? {
        val obj = data.opt(name)
        return when (obj) {
            is BigDecimal -> obj
            is Double -> BigDecimal(obj).roundedValue(decimals = decimals).stripTrailingZeros()
            is Number -> BigDecimal(obj.toDouble()).roundedValue(decimals = decimals).stripTrailingZeros()
            else -> null
        }
    }

    @Throws(JSONException::class)
    fun getDecimal(name: String, decimals: Int) = optDecimal(name, decimals = decimals)
        ?: throw JSONException("Expected double for " + name)

    fun optBoolean(name: String): Boolean? = data.opt(name) as? Boolean

    @Throws(JSONException::class)
    fun getBoolean(name: String): Boolean = optBoolean(name)
        ?: throw JSONException("Expected double for " + name)

    fun optString(name: String) = data.opt(name) as? String

    @Throws(JSONException::class)
    fun getString(name: String) = optString(name)
        ?: throw JSONException("Expected String for " + name)

    fun optDate(name: String): Date? {
        val obj = data.opt(name) as? String ?: return null
        return Dates.from(string = obj, format = Dates.defaultFormat, timeZone = Dates.UTC_TIME_ZONE)
    }

    @Throws(JSONException::class)
    fun getDate(name: String) = optDate(name) ?: throw JSONException("Expected Date for " + name)

    fun optTime(key: String): Time? {
        val value = data[key] as? String ?: return null
        val components = value.split(":").compactMap { it.toIntOrNull() }
        if (components.size != 3) {
            return null
        }
        return Time.of(hour = components[0], minute = components[1], second = components[2])
    }

    fun getTime(key: String): Time = optTime(key) ?: throw JSONException("Expected Time for " + key)

    fun optJSONObject(name: String): JSONObject? {
        if (data.isNull(name)) {
            return null
        }
        val obj = data.opt(name)
        if (obj is org.json.JSONObject) {
            return JSONObject(obj)
        }

        return null
    }

    @Throws(JSONException::class)
    fun getJSONObject(name: String) = optJSONObject(name)
        ?: throw JSONException("Expected JSONObject for " + name)

    fun optJSONArray(name: String): JSONArray? {
        val obj = data.opt(name)
        if (obj is org.json.JSONArray) {
            return JSONArray(obj)
        }

        return null
    }

    @Throws(JSONException::class)
    fun getJSONArray(name: String) = optJSONArray(name)
        ?: throw JSONException("Expected JSONArray for " + name)

    fun optStringArray(name: String): List<String>? {
        try {
            val array = getJSONArray(name)
            return array.map { item -> item as String }
        } catch (je: JSONException) {
            return null
        }
    }

    @Throws(JSONException::class)
    fun getStringArray(name: String) = optStringArray(name)
        ?: throw JSONException("Expected String Array for " + name)

    fun optIntArray(name: String): List<Int>? {
        try {
            val array = getJSONArray(name)
            return array.map { item -> item as Int }
        } catch (je: JSONException) {
            return null
        }
    }

    @Throws(JSONException::class)
    fun getIntArray(name: String) = optIntArray(name)
        ?: throw JSONException("Expected Int Array for " + name)

    fun optAnyArray(name: String): List<Any>? {
        try {
            val array = getJSONArray(name)
            return array.map { it }
        } catch (je: JSONException) {
            return null
        }
    }

    @Throws(JSONException::class)
    fun getAnyArray(name: String) = optAnyArray(name)
        ?: throw JSONException("Expected Any Array for " + name)

    @Throws(JSONException::class)
    fun <T> asMap(): Map<String, T> {
        val map = mutableMapOf<String, T>()
        val iterator = data.keys()
        while (iterator.hasNext()) {
            val key = iterator.next()
            if (data.isNull(key)) { continue }
            val obj = data.get(key)
            map[key] = when (obj) {
                is org.json.JSONArray -> JSONArray(obj).asList<Any>()
                is org.json.JSONObject -> JSONObject(obj).asMap<Any>()
                else -> obj
            } as T
        }

        return map
    }

    fun setAny(key: String, value: Any?) {
        when (value) {
            is Int -> setInt(key, value)
            is Boolean -> setBoolean(key, value)
            is Date -> setDate(key, value)
            is BigDecimal -> setDecimal(key, value)
            is Double -> setDouble(key, value)
            is String -> setString(key, value)
            is Time -> setTime(key, value)
            is List<*> -> setAnyArray(key, value as List<Any>)
            is Map<*, *> -> setObject(key, JSONObject(value as Map<String, Any>))
            else -> data.put(key, value)
        }
    }

    @Throws(JSONException::class)
    fun setInt(key: String, value: Int?) {
        data.put(key, value)
    }

    @Throws(JSONException::class)
    fun setDouble(key: String, value: Double?) {
        data.put(key, value)
    }

    @Throws(JSONException::class)
    fun setDecimal(key: String, value: BigDecimal?) {
        data.put(key, value)
    }

    @Throws(JSONException::class)
    fun setString(key: String, value: String?) {
        data.put(key, value)
    }

    @Throws(JSONException::class)
    fun setBoolean(key: String, value: Boolean?) {
        data.put(key, value)
    }

    @Throws(JSONException::class)
    fun setObject(key: String, value: JSONObject?) {
        if (value != null) {
            data.put(key, value.data)
        } else {
            data.remove(key)
        }
    }

    @Throws(JSONException::class)
    fun setAnyArray(key: String, value: List<Any>?) {
        if (value != null) {
            val array = JSONArray()
            value.forEach { array.addAny(it) }
            data.put(key, array.data)
        } else {
            data.remove(key)
        }
    }

    @Throws(JSONException::class)
    fun setArray(key: String, value: List<JSONObject>?) {
        if (value != null) {
            val array = JSONArray()
            array.addAll(value)
            data.put(key, array.data)
        } else {
            data.remove(key)
        }
    }

    @Throws(JSONException::class)
    fun setJSONArray(key: String, value: JSONArray?) {
        if (value != null) {
            data.put(key, value.data)
        } else {
            data.remove(key)
        }
    }

    @Throws(JSONException::class)
    fun setStringArray(key: String, value: List<String>?) {
        if (value != null) {
            val array = JSONArray()
            array.addAllString(value)
            data.put(key, array.data)
        } else {
            data.remove(key)
        }
    }

    @Throws(JSONException::class)
    fun setIntArray(key: String, value: List<Int>?) {
        if (value != null) {
            val array = JSONArray()
            array.addAllInt(value)
            data.put(key, array.data)
        } else {
            data.remove(key)
        }
    }

    @Throws(JSONException::class)
    fun setDate(key: String, value: Date?) {
        if (value == null) {
            data.remove(key)
            return
        }
        data.put(key, value.toString(format = Dates.defaultFormat, timeZone = Dates.UTC_TIME_ZONE))
    }

    fun setTime(key: String, value: Time?) {
        if (value == null) {
            data.remove(key)
            return
        }
        data.put(key, value.description)
    }

    fun toData() = toString().toByteArray()

    override fun toString() = data.toString().replace("\\/", "/")
}
